File: c/sliding_max_sum.c
#include <stdio.h>
#include <strings.h>
#include <stdlib.h>
#include <errno.h>
#include <err.h>
#include <limits.h>

int main(int argc, char** argv) {

    char in[sizeof(int)+1];
    int i, k, n;
    int *arr;

    printf("\nEnter the number of #s to sum at a time: " );
    fgets(&in, sizeof(int)+1, stdin);
    k = strtol(in, NULL, 10);


    do {
        printf("\nEnter the total number of #s being summed over: " );
        fgets(&in, sizeof(int)+1, stdin);
        n = strtol(in, NULL, 10);
    } while (n < k);

    if (!(arr = (int*) malloc(sizeof(int)*n))) {
        err(errno, NULL);
    }

    int max = INT_MIN; 
    int sum = 0;
    int fpos = 0;
    int count = 0;

    for (i=0;i < n;i++) {
        printf("\nEnter the next number to sum over (%i left): ", (n - i));
        fgets(&in, sizeof(int)+1, stdin);
        arr[i] = strtol(in, NULL, 10);

        if (count < k) {
            sum += arr[i];
            count++;
            continue;
        }

        if (max < sum)
            max = sum;
        
        sum -= arr[fpos];
        sum += arr[i];
        fpos++;
    }

    free(arr);

    printf ("Maximum sum over %i numbers is %i\n", k, max);
}

File: py/sliding_max_sum.py
import sys

arr= [1, 4, 2, 10, 23, 3, 1, 0, 20]
k = 4

count = 0
max_int = -(sys.maxsize)

sum_k = 0

first_ix = 0

ix = 0

while ix < len(arr):
    if count < k:
        count += 1
        sum_k += arr[ix]
        ix += 1
        continue

    if sum_k > max_int:
        max_int = sum_k

    sum_k -= arr[first_ix]
    sum_k += arr[ix]
    first_ix += 1
    ix += 1

if max_int < sum_k:
    max_int = sum_k

print("Max sum of {} elements is {}".format(k, max_int))


